### DISCUSSION TITLE
Implement Catmull-Rom 4-Tap as a Scaling 3D Option

### DISCUSSION THREAD

**Discussion by cybereality** [godotengine/godot-proposals#6737] [Platforms]
I have found an inexpensive way to upscale an image that is slightly better than linear. It is not on the level of FSR, however it can help improve the sharpness and image quality on low-end devices, specifically mobile and web. The technique and code is shown on this page, with a shadertoy demo so you can see how it works.

https://www.decarpentier.nl/2d-catmull-rom-in-4-samples
https://www.shadertoy.com/view/4tyGDD

It does require a shader, so there is some cost above built-in linear, but should be much less than FSR, particularly on the low-end.

I have implemented this in my own OpenGL demo to demonstrate the quality. This is 540P upscaled to 1080P. It still looks low-res, of course, but you can notice on the leather pattern and on the ribbed dial that it retains detail better. The text is also slightly more clear and overall does not look as blurry and muddy. Note, this is on the lowest setting that is viable (50%) on something like 70% it looks close to native with just some additional aliasing on edges. Particularly for smaller screens on mobile, the loss of quality is not that bad but the performance boost is significant.

![Catmull_1](https://user-images.githubusercontent.com/3398643/233774199-ed8d04be-bd6e-4bdc-9ead-bdc581234d51.jpg)

![Catmull_2](https://user-images.githubusercontent.com/3398643/233774205-e2d61804-0099-4621-b671-750e3eb4e366.jpg)

![Catmull_3](https://user-images.githubusercontent.com/3398643/233774207-c1a3fb13-5ac2-49f9-8f63-a560fed18f21.jpg)

**Comment by cybereality** [comment]
Here is a 1080P shot with 70% render scale, which is think looks pretty close to native (aside from the aliasing).

![Catmull_4](https://user-images.githubusercontent.com/3398643/233781067-ad9e0494-aa27-437a-8c32-861303713828.jpg)

**Comment by cybereality** [comment]
This is 720P upscaled to 1440P but with an FXAA pass on the original buffer (before scaling).

![degine_catmull_fxaa](https://user-images.githubusercontent.com/3398643/233785574-dc0c89c0-1ebe-4990-98f0-4cc2fd262bf0.jpg)

**Comment by Calinou** [comment]
If performance is a priority, I would just focus on [reimplementing CAS](https://github.com/godotengine/godot/pull/47401) and using that with bilinear scaling. CAS is very cheap and easily runs in 0.1 ms of GPU time or less. [FSR Lite](https://github.com/godotengine/godot-proposals/issues/3786) is also an option (more expensive than CAS, but less than full-blown FSR 1.0).

**Comment by cybereality** [comment]
I would have to do some performance tests as I'm just ballparking it here. However, I do think FXAA + Catmull-Rom looks significantly better than CAS. CAS suffers from a lot of dark edges and sometimes can look like a sobel filter, whereas Catmull-Rom looks closer to native (as you can see in the last screenshot I posted). That said, I can take a look at CAS and FSR Lite and see if they are better options.
